Improvement: add an option to watch messaging. Add .idea to .gitignore

EugeneHlushko opened the pull request:
<!-- Thanks for submitting a pull request! Please provide enough information so that others can review your pull request. -->

**What kind of change does this PR introduce?**
<!-- E.g. a bugfix, feature, refactoring, build related change, etc… -->
Adds `infoVerbosity` option which is set to `info` by default. In order for user to be able to control the amount of extra messaging output. 

Provides clear indication of when compilation had started and when it had finished if such verbosity is requested

Fixes webpack/webpack#6051 and webpack/webpack#5895

**Did you add tests for your changes?**
Yes
Also added PR for json schema validation of the option: webpack/webpack#6332
UPD: we wont add these to webpkac/webpack's schema, instead we can create our own in webpack-cli

**If relevant, did you update the documentation?**
*Not yet had been approved*

**Summary**
<!-- Explain the **motivation** for making this change. What existing problem does the pull request solve? -->
<!-- Try to link to an open issue for more information. -->
See webpack/webpack#6051 and webpack/webpack#5895

**Does this PR introduce a breaking change?**
<!-- If this PR introduces a breaking change, please describe the impact and a migration path for existing applications. -->
no

**Other information**

EugeneHlushko commented:
Hi @sokra, here's is migrated pr to webpack-cli. Couple of tests are now failing due to message being off by default. Should i turn it on by default or should i change the tests to not expect the messages when the param is not there?

ev1stensberg commented:
@EugeneHlushko update the tests, would be convenient if you add a way to test this locally, not just via tests too / screenshot 👍

EugeneHlushko commented:
Hi @ev1stensberg 

i've updated the tests with config to expect the string and default ones to not expect it. Do you think that is enough or should i just add this option to all old tests and create a separate test case to see that string?

EugeneHlushko commented:
Any update on this one @ev1stensberg ?

ev1stensberg commented:
Is there an issue on this? If not , you'd have to open one and @sokra must approve the feature

sokra commented:
See also https://github.com/webpack/webpack/issues/766

I guess it should be on by default.

See also https://github.com/webpack/webpack/issues/5895

I guess it would a a good idea to combine these features into something more general like `info-verbosity`...

EugeneHlushko commented:
Got your point @sokra 
`info-verbosity` sounds more like a level setting rather than a bool. In that case i would make it: `verbose` and `silent` in this PR, and in future we could always add third option and so on.

And by default it would be set to `verbose` per @ev1stensberg 's request.

If you are ok with that the plan for me would be adding that option within the scope of this PR, adding additional test case for `silent` mode. And the next PR would then cover more messages like `Build started...` and `Build finished` in the watch mode.

Please comment or approve the course.

P.S. Updating the description with permalinks to issues

ev1stensberg commented:
@EugeneHlushko @sokra  I actually think that we should try to align this with #22 , as this is the same issue?

EugeneHlushko commented:
Not exactly the same but similar. If you read the issues, its more about being able to not see extra text except for the compilation output, and another one wants to see clear messages on `start` and `end` of the compilation process. But we indeed can also respect this flag for `hash` and `version` etc. Can include it into the scope of current PR, what you think?

ev1stensberg commented:
I think you should consult with @sokra, it would be nice if we'd combine these features into one alias

EugeneHlushko commented:
Ok lets see what @sokra can say on this thread after what i've wrote yesterday: https://github.com/webpack/webpack-cli/pull/200#issuecomment-351736138

EugeneHlushko commented:
Hi @sokra , @ev1stensberg 
As this isnt getting additional comments i've implemented the last suggested solution and added a new test case. Please review

EugeneHlushko commented:
Any updates on this PR? the two issues opened are not getting closed if this is freezed. @sokra @ev1stensberg

sokra commented:
> info-verbosity sounds more like a level setting rather than a bool.

yep, in my opinion it should be a level settings.

I would default it to "info".

In "verbose" you could add the messages for start, end invalid...

EugeneHlushko commented:
Hi @ev1stensberg and @sokra 
I did update the PR accordingly

EugeneHlushko commented:
Should i rebase or is this still under discussion @sokra ?

sokra commented on bin/webpack.js:
```diff
@@ -378,6 +384,10 @@ yargs.parse(process.argv.slice(2), (err, argv, output) => {
 		});
 
+		ifArg("info-verbosity", function(value) {
+			outputOptions.infoVerbosity = value;
```
You don't have to put this into `outputOptions`. A local variable would be fine.

ev1stensberg commented:
Could you rebase?  @EugeneHlushko

EugeneHlushko commented:
Yup, on it. I will leave it in outputOptions as i thought to use it in the plugins too, but if you insist i can change to local var

EugeneHlushko commented:
done @ev1stensberg

ev1stensberg commented:
Hi @EugeneHlushko . Thanks! Will review tomorrow!

EugeneHlushko commented:
@ev1stensberg that one stretched out, any updates?

ev1stensberg commented:
Hey, sorry, forgot, hectic these days. I'll review tomorrow @EugeneHlushko

ev1stensberg commented on bin/webpack.js:
```diff
@@ -183,6 +183,12 @@
+			type: "string",
+			default: "info",
+			group: DISPLAY_GROUP,
+			describe: "Controls the output of lifecycle messaging e.g. Started watching files... (verbose, info, none)"
```
What's the difference between `info` and `verbose` ? Seems to do the same thing.

EugeneHlushko replied:
Well within this PR yes, but the next one should add two more messages they will be verbose only (see linked issues)

ev1stensberg replied:
Could you make it in this PR as well? Ask any of the maintainers if you got any questions!

EugeneHlushko replied:
Probably it will stretch out for another month in that case, any good reason for that?

ev1stensberg replied:
Leaving code unfinished is a bad habit 💃

EugeneHlushko replied:
But if you promise to respond quickly im on it! :)

ev1stensberg replied:
Pinky Promise 😆

ev1stensberg commented on bin/webpack.js:
```diff
@@ -382,6 +388,10 @@
 			});
 
+			ifArg("info-verbosity", function(value) {
+				outputOptions.infoVerbosity = value;
```
If `info-verbosity` is not a string, i.e boolean, it will still output. Would be nice with an error that will match against `webpack-schema`. https://github.com/webpack/webpack/blob/master/schemas/webpackOptionsSchema.json

You should send a PR to add `info-verbosity` and link it to this.

EugeneHlushko replied:
I will modify within this scope (the schema), compeltely forgot about it

EugeneHlushko replied:
schema added and pr's linked in description section

ev1stensberg replied:
Cool beans 🐻

ev1stensberg commented on bin/webpack.js:
```diff
@@ -448,7 +458,8 @@
 				}
 				compiler.watch(watchOptions, compilerCallback);
-				console.log("\nWebpack is watching the files…\n");
+				if (outputOptions.infoVerbosity !== "none")
```
See above

EugeneHlushko commented:
Thank you @ev1stensberg 👍 
I have answered the level purpose question in the code comment and schema will be updated later on today, i will ping you when its done.

EugeneHlushko commented:
@ev1stensberg feature for webpack/webpack#5895 and test coverage added

EugeneHlushko commented:
I've also added temporary validation, we can deprecate it as soon as jsonschema is implemented.

EugeneHlushko commented:
But i did add that, or do you mean anything else?
Here:
https://github.com/webpack/webpack-cli/pull/200/files#diff-877a99a662829398e8136ee938cf1c93R443
and here:
https://github.com/webpack/webpack-cli/pull/200/files#diff-877a99a662829398e8136ee938cf1c93R489

EugeneHlushko commented:
Hi @ev1stensberg 
Forgot to mention you in the last comment.
Verbose is already added in previous commits:
https://github.com/webpack/webpack-cli/pull/200/files#diff-877a99a662829398e8136ee938cf1c93R443

ev1stensberg commented:
Let me take another look later tonight

EugeneHlushko commented:
Sure thing mate, looking forward to feedback if any @ev1stensberg

ev1stensberg commented on bin/webpack.js:
```diff
@@ -399,6 +405,12 @@
 
+			ifArg("info-verbosity", function(value) {
+				if (typeof value !== "string" || !["none", "info", "verbose"].includes(value))
+					throw new Error("Invalid configuration object. \n configuration['info-verbosity'] should be one of these:\n \"none\" | \"info\" | \"verbose\"");
```
I think we've got a error helper for this

EugeneHlushko replied:
I maybe misunderstood, but couldnt find the usage in codebase of error helper, could you please point me into the right direction here? All i've found is `bin/errorHelpers.js` which basically works with existing error but doesnt throw one for me

ev1stensberg replied:
Hm, okay, so if you look in the function inside the bin folder that I believe is called convert-argv, there's a processing of the options, and if it's not matching the schema, it throws. At least that is what I think it's doing under the hood. I'm not on a computer right now, so let me get back to you in a bit

EugeneHlushko replied:
yeah but seems it doesnt support enum which could substitue `![].includes` call as jsonschema does?

ev1stensberg replied:
It should only be a string

EugeneHlushko commented on bin/webpack.js:
```diff
@@ -399,6 +405,12 @@
 			});
 
+			ifArg("info-verbosity", function(value) {
+				if (!["none", "info", "verbose"].includes(value))
```
@ev1stensberg can we keep this check until schema lands so there is no misuse?

ev1stensberg replied:
Could you create an issue for it? Mergeable after

EugeneHlushko replied:
done #249 @ev1stensberg

ev1stensberg commented:
Do you mind adding `webpack-dev-server` as the same dependency as `webpack`? Seems to get linting errors

EugeneHlushko commented:
@ev1stensberg added `webpack-dev-server` to `allowModules` section
